### PR TITLE
accommodate list-columns in `grid`

### LINKED ISSUE #625: tuning with list-columns in `grid`
I want to customize objective function. 

For example, what I want to implement now is a loss function called Focal loss, which can be used when there is a class imbalance.
[Focal Loss](https://maxhalford.github.io/blog/lightgbm-focal-loss/)

Because focal loss is a generalization of cross-entropy, setting certain hyperparameters will result in exactly the same result as CE.

``` r
library(tidymodels)
library(xgboost)
#> 
#> Attaching package: 'xgboost'
#> The following object is masked from 'package:dplyr':
#> 
#>     slice

data(agaricus.train, package = "xgboost")
data(agaricus.test, package = "xgboost")

dtrain <- with(agaricus.train, xgb.DMatrix(data, label = label, nthread = 2))
dtest <- with(agaricus.test, xgb.DMatrix(data, label = label, nthread = 2))
watchlist <- list(train = dtrain, eval = dtest)

# original cross entropy loss
logregobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  return(list(grad = grad, hess = hess))
}

# focal_loss --------------------------------------------------------------

focal_loss <- function(preds, dtrain,alpha=0.5,focal_gamma=0) {
  labels <- getinfo(dtrain, "label")
  preds <- 1 / (1 + exp(-preds))
  
  p<- preds
  y <- labels
  
  grad <- (y*alpha*(1-p)^focal_gamma*(focal_gamma*log(p)*p-(1-p))-
             (1-y)*(1-alpha)*p^(focal_gamma)*(focal_gamma*log(1-p)*(1-p)-p))
  
  
  
  du <- y*alpha*(1-p)^(focal_gamma-1)*(log(p)*(-focal_gamma^2 *p + (1-p)*focal_gamma)+ 2*focal_gamma*(1-p)+(1-p))
  dv <- -(1-y)*(1-alpha)*p^(focal_gamma-1)*(log(1-p)*(focal_gamma^2*(1-p)-p*focal_gamma)-2*focal_gamma*p-p)
  
  hess <- (du+dv)*p*(1-p)
  
  return(list(grad = grad, hess = hess))
}


param <- list(max_depth = 2, eta = 1, nthread = 2,
              objective = "binary:logistic", eval_metric = "auc")
bst <- xgb.train(param, dtrain, nrounds = 5, watchlist)
#> [1]  train-auc:0.958228  eval-auc:0.960373 
#> [2]  train-auc:0.981413  eval-auc:0.979930 
#> [3]  train-auc:0.997070  eval-auc:0.998518 
#> [4]  train-auc:0.998757  eval-auc:0.998943 
#> [5]  train-auc:0.999298  eval-auc:0.999830
#> [1]  train-auc:0.958228  eval-auc:0.960373 
#> [2]  train-auc:0.981413  eval-auc:0.979930

bst$params$objective
#> [1] "binary:logistic"
#> [1] "binary:logistic"

param$objective <- logregobj

bst <- xgb.train(param, dtrain, nrounds = 5, watchlist)
#> [1]  train-auc:0.958228  eval-auc:0.960373 
#> [2]  train-auc:0.981413  eval-auc:0.979930 
#> [3]  train-auc:0.997070  eval-auc:0.998518 
#> [4]  train-auc:0.998757  eval-auc:0.998943 
#> [5]  train-auc:0.998120  eval-auc:0.999830
#> [1]  train-auc:0.958228  eval-auc:0.960373 
#> [2]  train-auc:0.981413  eval-auc:0.979930

param$objective <- focal_loss# alpha = 0.5, gamma= 0 -> same result! 

bst <- xgb.train(param, dtrain, nrounds = 5, watchlist)
#> [1]  train-auc:0.958228  eval-auc:0.960373 
#> [2]  train-auc:0.981413  eval-auc:0.979930 
#> [3]  train-auc:0.997070  eval-auc:0.998518 
#> [4]  train-auc:0.998166  eval-auc:0.998943 
#> [5]  train-auc:0.998823  eval-auc:0.999830

```
# I want to tuning alpha , focal_gamma

``` r
param$objective <- partial(focal_loss, alpha = 0.3, focal_gamma = 0)

bst <- xgb.train(param, dtrain, nrounds = 5, watchlist)
#> [1]  train-auc:0.979337  eval-auc:0.980196 
#> [2]  train-auc:0.992593  eval-auc:0.993159 
#> [3]  train-auc:0.999934  eval-auc:0.999917 
#> [4]  train-auc:0.999978  eval-auc:0.999972 
#> [5]  train-auc:0.999978  eval-auc:0.999972
```

<sup>Created on 2023-02-28 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I checked that it works well on the existing xgb.train, but I don't know how to apply the tune function. 
From now on, the code below is the way I tried.

``` r
library(tidymodels)
library(xgboost)
#> 
#> Attaching package: 'xgboost'
#> The following object is masked from 'package:dplyr':
#> 
#>     slice
library(scales)
library(dials)



alpha <- function(range = c(0,1), trans = NULL) {
  new_quant_param(
    type = "double",
    range = range,
    inclusive = c(TRUE, TRUE),
    trans = trans,
    label = c(num_initial_terms = "# Initial alpha"),
    finalize = NULL
  )
}


focal_gamma <- function(range = c(0, 5), trans = NULL) {
  new_quant_param(
    type = "double",
    range = range,
    inclusive = c(TRUE, TRUE),
    trans = trans,
    label = c(num_initial_terms = "# Initial gamma"),
    finalize = NULL
  )
}


data<- two_class_dat |> 
  rename(y=Class)

set.seed(100)
splits<- initial_split(data,prop = 0.8,strata = y)
train_data <- training(splits)
test_data <- testing(splits)
resamples<- vfold_cv(data = train_data,v = 5,strata = y)

xgb_model <- boost_tree( mode = "classification",
                               tree_depth     =tune(),
                               trees          =tune()) |> 
  set_engine(engine = "xgboost" ,
             objective = partial(focal_loss,
                                 focal_gamma=tune())) # 
#I wanted to tune the two hyperparameters together at first, but due to the error message, I am aiming to tune one first.

#>Error: Only one tunable value is currently allowed per argument.
#>The current argument has: `partial(focal_loss, focal_gamma = tune(), alpha = tune())`.



xgb_model |> translate()
#> Boosted Tree Model Specification (classification)
#> 
#> Main Arguments:
#>   trees = tune()
#>   tree_depth = tune()
#> 
#> Engine-Specific Arguments:
#>   objective = partial(focal_loss, focal_gamma = tune())
#> 
#> Computational engine: xgboost 
#> 
#> Model fit template:
#> parsnip::xgb_train(x = missing_arg(), y = missing_arg(), weights = missing_arg(), 
#>     nrounds = tune(), max_depth = tune(), objective = partial(focal_loss, 
#>         focal_gamma = tune()), nthread = 1, verbose = 0)



rec_base<- train_data %>% 
  recipe(y~.) |> 
  #step_mutate_at(all_numeric_predictors(), fn = list(orig = ~.)) %>%
  step_normalize(all_predictors(), -all_outcomes()) 


xgb_workflow <- 
  workflow() %>% 
  add_recipe(rec_base) %>% 
  add_model(xgb_model)


xgb_workflow %>%
  extract_parameter_set_dials()
#> Collection of 3 parameters for tuning
#> 
#>  identifier       type    object
#>       trees      trees nparam[+]
#>  tree_depth tree_depth nparam[+]
#>   objective  objective   missing
#> The parameter `objective` needs a `param` object. 
#> See `vignette('dials')` to learn more.
```

<sup>Created on 2023-02-28 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

The `extract_parameter_set_dials()` function does not recognize the `focal_gamma` argument.
How I can tuning objective function?
Ultimately, I want to tune not only one parameter but also several parameters together.

### PR DESCRIPTION
**Pull request by simonpcouch**
Closes #625—more discussion there.

Here's the reprex from that issue, with this PR:

``` r
library(tidymodels)

# focal_loss --------------------------------------------------------------
focal_loss <- function(preds, dtrain, alpha = 0.5,focal_gamma = 0) {
  labels <- getinfo(dtrain, "label")
  preds <- 1 / (1 + exp(-preds))
  
  p<- preds
  y <- labels
  
  grad <- (y*alpha*(1-p)^focal_gamma*(focal_gamma*log(p)*p-(1-p))-
             (1-y)*(1-alpha)*p^(focal_gamma)*(focal_gamma*log(1-p)*(1-p)-p))
  
  
  
  du <- y*alpha*(1-p)^(focal_gamma-1)*(log(p)*(-focal_gamma^2 *p + (1-p)*focal_gamma)+ 2*focal_gamma*(1-p)+(1-p))
  dv <- -(1-y)*(1-alpha)*p^(focal_gamma-1)*(log(1-p)*(focal_gamma^2*(1-p)-p*focal_gamma)-2*focal_gamma*p-p)
  
  hess <- (du+dv)*p*(1-p)
  
  return(list(grad = grad, hess = hess))
}

# data setup ----------------------------------------------------------------
data <- two_class_dat %>% 
  rename(y = Class)

set.seed(100)

splits <- initial_split(data,prop = 0.8, strata = y)
train_data <- training(splits)
test_data <- testing(splits)
resamples <- vfold_cv(data = train_data,v = 5,strata = y)

# specifications ----------------------------------------------------
xgb_spec <- 
  boost_tree(mode = "classification", tree_depth = tune(), trees = tune()) %>% 
  # note that i just set `objective = tune()` here
  set_engine(object = ., engine = "xgboost", objective = tune())

xgb_recipe <- train_data %>% 
  recipe(y ~ .) |> 
  #step_mutate_at(all_numeric_predictors(), fn = list(orig = ~.)) %>%
  step_normalize(all_predictors(), -all_outcomes()) 

xgb_workflow <- 
  workflow() %>% 
  add_recipe(xgb_recipe) %>% 
  add_model(xgb_spec)

# grid setup ------------------------------------------------------------------
partial_grid <-
  expand_grid(
    alpha = seq(0, 1, length.out = 3),
    focal_gamma = seq(0, 5, length.out = 3)
  )

partial_grid$partials <-
  map2(
    partial_grid$alpha,
    partial_grid$focal_gamma,
    ~partial(focal_loss, alpha = !!.x, focal_gamma = !!.y)
  )
    
xgb_grid <-
  extract_parameter_set_dials(xgb_spec) %>%
  filter(id != "objective") %>%
  grid_latin_hypercube(size = 9) %>%
  bind_cols(partial_grid %>% select(objective = partials))

# tune! -----------------------------------------------------------------------
res <- tune_grid(xgb_workflow, resamples = resamples, grid = xgb_grid)

res
#> # Tuning results
#> # 5-fold cross-validation using stratification 
#> # A tibble: 5 × 4
#>   splits            id    .metrics          .notes          
#>   <list>            <chr> <list>            <list>          
#> 1 <split [505/127]> Fold1 <tibble [18 × 7]> <tibble [0 × 3]>
#> 2 <split [505/127]> Fold2 <tibble [18 × 7]> <tibble [0 × 3]>
#> 3 <split [505/127]> Fold3 <tibble [18 × 7]> <tibble [0 × 3]>
#> 4 <split [506/126]> Fold4 <tibble [18 × 7]> <tibble [0 × 3]>
#> 5 <split [507/125]> Fold5 <tibble [18 × 7]> <tibble [0 × 3]>
```

<sup>Created on 2023-03-09 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by simonpcouch** [comment]
Tagging @hfrick—I'm not sure if this is a band-aid for not being able to tune the arguments to that partial properly or a sound fix to an actual issue. Would love to hear if you think there may be a better (existing?) approach to solving that problem. :)

**Comment by SHo-JANG** [comment]
With this version, the verbose option does not work!

**Comment by simonpcouch** [comment]
> With this version, the verbose option does not work!

Thanks for the follow-up! I'd be surprised if this were to have any effect on the verbosity settings, though there are other changes in the dev version that might be relevant. Do you see the same issues with the version on `pak::pak("tidymodels/tune")` installed, and if so, could you open a separate issue with a reprex?

**Review by hfrick — APPROVED** [review]
Here is a quick recap of our conversation as a paper trail: 

I think this change is in the spirit of tune since it essentially allows you to tune a placeholder for functions, and I don't see why we wouldn't be okay with that. You would need to put together your own custom grid as you did in your reprex, I don't see a quick way via dials for that.

**Review by topepo — APPROVED** [review]

**Comment by SHo-JANG** [comment]
I want to search parameter like this.

```r
extract_parameter_set_dials(xgb_spec) %>%
  filter(id != "objective") %>%
  grid_latin_hypercube(size = 9) %>%
  mutate(obj_list = list(partial_grid %>% select(objective = partials))) %>%
  unnest(obj_list)
```
But this method is too expensive.. 
I hope to be able to use the `tune_bayes` function.
